akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala

```diff
@@ -315,23 +294,60 @@ abstract class GraphStageLogic private[stream] (inCount: Int, outCount: Int) {
     }
   }
 
-  private def onInternalTimer(scheduled: Scheduled): Unit = {
-    val Id = scheduled.timerId
-    keyToTimers.get(scheduled.timerKey) match {
-      case Some(Timer(Id, _)) ⇒
-        if (!scheduled.repeating) keyToTimers -= scheduled.timerKey
-        onTimer(scheduled.timerKey)
-      case _ ⇒
-    }
+  // Internal hooks to avoid reliance on user calling super in preStart
+  protected[stream] def beforePreStart(): Unit = ()
+
+  // Internal hooks to avoid reliance on user calling super in postStop
+  protected[stream] def afterPostStop(): Unit = ()
+
+  /**
+   * Invoked before any external events are processed, at the startup of the stage.
+   */
+  def preStart(): Unit = ()
+
+  /**
+   * Invoked after processing of external events stopped because the stage is about to stop or fail.
+   */
+  def postStop(): Unit = ()
+}
+
+abstract class TimerGraphStageLogic(_shape: Shape) extends GraphStageLogic(_shape) {
```

Comment:
I pushed this down as a subclass, since it hooks into the beforePreStart/afterPostStop hooks, otherwise it could be a separate support class that you just compose with a GraphStageLogic. If anyone has a good solution what to do here (trait composition won't work for Java) please share.